src/txmempool.cpp

```diff
@@ -2009,6 +2071,72 @@ size_t CTxMemPool::DynamicMemoryUsage() const {
           cachedInnerUsage);
 }
 
+bool CTxMemPool::trimToSize(const CMemPoolEntry* entry, size_t max_size) {
+    LOCK(cs);
+    size_t new_entry_usage = entry ? entry->GetSize() : 0;
+    assert(new_entry_usage < max_size);
+    size_t current_usage = totalTxSize + totalCertificateSize;
+    // If the mempool size is within the limits, do nothing and return immediatly.
+    if (current_usage + new_entry_usage <= max_size) return true;
+
+    const CCertificateMemPoolEntry* certificate = dynamic_cast<const CCertificateMemPoolEntry*>(entry);
+    if (certificate && totalTxSize > m_max_size / 2) {
+        LogPrint("mempool", "%s():%d - Trying to remove transactions to make room for certificate (tot_tx: %zu, max: %zu)\n", __func__, __LINE__, totalTxSize, m_max_size / 2);
+        trimToSize(nullptr, max_size - new_entry_usage);
+        current_usage = totalTxSize + totalCertificateSize;
+        if (current_usage + new_entry_usage <= m_max_size) return true;
+    }
+
+    // we must either reject incoming tx, or remove something
+    int64_t size_to_be_removed = current_usage + new_entry_usage - max_size;
```

Comment:
Please, change these to std::unordered_map. It has better operational complexity.

Reply:
I agree, done with the latest set of updates.